File: examples/Simple/client/rb/AdvancedStatic/SimpleClient.rb
#!/usr/bin/env ruby 

$:.push('../gen-rb')

require 'thrift'
require 'simple_service'

def check(data_out, expected, size)
    status = 0
    for i in 0..(size-1)
        if (data_out[i] != expected[i])
            puts data_out[i]
            puts expected[i]
            status = 1
        end
    end
    return status
end

def SimpleCPU(size, data_in, data_out)
    for i in 0..(size-1)
        data_out[i] = (data_in[i] * data_in[i] + data_in[i])
    end
end

begin
    include Com::Maxeler::Simple

    port = 9090

    # Make socket
    transport = Thrift::BufferedTransport.new(Thrift::Socket.new('localhost', port))

    # Wrap in a protocol
    protocol = Thrift::BinaryProtocol.new(transport)

    # Create a client to use the protocol encoder
    client = SimpleService::Client.new(protocol)

    # Connect!
    transport.open()
 
    size = 384        

    # Generate input
    data_in = Array.new(size)

    for i in 0..(size-1)
        data_in[i] = i + 1
    end

    # Initialize maxfile
    max_file = client.Simple_init()

    # Load DFE
    max_engine = client.max_load(max_file, '*')

    # Allocate and send input streams to server
    address_data_in = client.malloc_float(size)
    client.send_data_float(address_data_in, data_in)

    # Allocate memory for output stream on server
    address_data_out = client.malloc_float(size)

    puts "Running DFE."
    actions = Simple_actions_t_struct.new()
    actions.param_N = size
    actions.instream_x = address_data_in
    actions.outstream_y = address_data_out
    address_actions = client.send_Simple_actions_t(actions)
    client.Simple_run(max_engine, address_actions)

    # Unload DFE
    client.max_unload(max_engine)

    # Get output stream from server
    data_out = client.receive_data_float(address_data_out, size)

    # Free allocated memory for streams on server
    client.free(address_data_in)
    client.free(address_data_out)

    # Free allocated maxfile data
    client.Simple_free()

    # Close!
    transport.close()

    # Checking results
    expected = Array.new(size)
    SimpleCPU(size, data_in, expected)

    status = check(data_out, expected, size)

    if (status == 1)
        puts "Test failed."
        Kernel.exit(-1)
    else
        puts "Test passed!"
    end


rescue Thrift::Exception => thrift_exception
    puts 'Thrift::Exception: ', thrift_exception.message, "\n"
    Kernel.exit(-1)
end

File: examples/Simple/client/perl/AdvancedStatic/SimpleClient.pl
#!/usr/bin/perl -w

use strict;
use warnings;

use lib '../gen-perl';

use Thrift;
use Thrift::BinaryProtocol;
use Thrift::Socket;
use Thrift::BufferedTransport;

use com::maxeler::Simple::SimpleService;
use com::maxeler::Simple::Types;

use Data::Dumper;

use Time::HiRes qw/ time /;

sub check {
    my($dataOutDFE, $dataOutCPU, $size) = @_;
    my $status = 0;

    for(my $i = 0; $i < $size; $i++) {
        if($$dataOutDFE[$i] != $$dataOutCPU[$i]) {
            print "Output data @ ".$i." = ".$$dataOutDFE[$i]." (expected ".$$dataOutCPU[$i].")\r\n";
            $status++;
        }
    }
    return $status;    
}

sub SimpleCPU {
    my ($size, $dataIn) = @_;
    my @dataOut = ();

    for (my $i = 0; $i < $size; $i++) {
        $dataOut[$i] = $$dataIn[$i] * $$dataIn[$i] + $$dataIn[$i];
    }

    return \@dataOut;
}

sub SimpleDFE {
    my ($size, $dataIn) = @_;
    my @dataOut = ();
    eval {
	my $startTime = time;
	
        # Make socket
        my $socket    = new Thrift::Socket('localhost', 9090);

        # Buffering is critical. Raw sockets are very slow
        my $transport = new Thrift::BufferedTransport($socket);

        # Wrap in a protocol
        my $protocol  = new Thrift::BinaryProtocol($transport);

        # Create a client to use the protocol encoder
        my $client    = new com::maxeler::Simple::SimpleServiceClient($protocol);

	print "Creating a client:\t\t\t\t", (time - $startTime), "s\n";

        # Connect!
	$startTime = time;
        $transport->open();
	print "Opening connection:\t\t\t\t", (time - $startTime), "s\n";

        # Initialize maxfile
	$startTime = time;
        my $max_file = $client->Simple_init();
	print "Initializing maxfile:\t\t\t\t", (time - $startTime), "s\n";

        # Load DFE
	$startTime = time;
        my $engine = $client->max_load($max_file, '*');  
	print "Loading DFE:\t\t\t\t\t", (time - $startTime), "s\n";
            
        # Allocate and send input streams to server
	$startTime = time;
        my $address_dataIn = $client->malloc_float($size);
        $client->send_data_float($address_dataIn, $dataIn);
	print "Sending input data:\t\t\t\t", (time - $startTime), "s\n";

        # Allocate memory for output stream on server
	$startTime = time;
        my $address_dataOut = $client->malloc_float($size);
	print "Allocating memory for output stream on server:\t", (time - $startTime), "s\n";
	
	# Action default        
        my $action = new com::maxeler::Simple::Simple_actions_t_struct();
        $action->{param_N} = $size;
        $action->{instream_x} = $address_dataIn;
        $action->{outstream_y}  = $address_dataOut;
        my $address_action = $client->send_Simple_actions_t($action);
        $client->Simple_run($engine, $address_action);
	print "Simple time:\t\t\t\t\t", (time - $startTime), "s\n";

        # Unload DFE
	$startTime = time;
        $client->max_unload($engine);
	print "Unloading DFE:\t\t\t\t\t", (time - $startTime), "s\n";

        # Get output stream from server
	$startTime = time;
        my $dataOut_ref = $client->receive_data_float($address_dataOut, $size);
        @dataOut = @$dataOut_ref;
	print "Getting output stream:\t(size = ", ($size * 32), " bit)\t", (time - $startTime), "s\n";

        # Free allocated memory for streams on server
	$startTime = time;
        $client->free($address_dataIn);
        $client->free($address_dataOut);
	print "Freeing allocated memory for streams on server:\t", (time - $startTime), "s\n";

        # Free allocated maxfile data
	$startTime = time;
        $client->Simple_free();
	print "Freeing allocated maxfile data:\t\t\t", (time - $startTime), "s\n";
        
	$startTime = time;
        $transport->close();
	print "Closing connection:\t\t\t\t", (time - $startTime), "s\n";

    }; if($@){
        warn(Dumper($@));
    }

    return \@dataOut;
}

my $startTime = time;
my $size = 1024;
my @dataIn = ();
my @dataOutCPU = ();
my @dataOutDFE = ();

for(my $i = 0; $i < $size; $i++) {
    $dataIn[$i] = $i + 1;
}
print "Generating input data:\t\t\t\t", (time - $startTime), "s\n";

# DFE Output
$startTime = time;
my $dataOutDFE_ref = SimpleDFE($size, \@dataIn);
print "DFE simple total time:\t\t\t\t", (time - $startTime), "s\n";

# CPU Output
$startTime = time;
my $dataOutCPU_ref = SimpleCPU($size, \@dataIn);
print "CPU simple total time:\t\t\t\t", (time - $startTime), "s\n";

@dataOutCPU = @$dataOutCPU_ref;
@dataOutDFE = @$dataOutDFE_ref;

# Checking results
$startTime = time;
my $status = check(\@dataOutDFE, \@dataOutCPU, $size);
print "Checking results:\t\t\t\t", (time - $startTime), "s\n";

if ($status == 0) {
    print "Test successful!\n";
} else {
        print "Test failed ", $status, " times!\n";
        exit -1;
}
